feat (fix?): a page tree with no page of its own lands on its first page (#370)
universal-ember/kolay: docs-app/src/templates/migrations/upgrading-from-5x.gjs.md

To assist in this migration, `<PageNav />` will `assert` in development when it is still given a `:collection` block to remind you to migrate to `:section`. The assertion is stripped from production builds.

`Node` is `Page | PageTree`. The `Runtime/utilities/collection-utils` page is now `page-tree-utils`, with a redirect from the old URL.

> ## A folder's index page is whichever page its URL goes to

Author: ```suggestion
## A folder's "index" page is now equivalent to the "landing page", even if no explicit "index" file exists
```

Author: Taken and then superseded within the same review. I applied it, and then your later point — that "landing page" is a confusing concept we should stop naming — meant the heading could not keep the word either.

It is `## A folder's index page falls back to its first page`, which says the change without introducing a second term to define.

*Drafted by Claude, reviewed before posting.*

> In 5.x a folder had an index page only if you wrote one named `index`. It is now the page named `index` when there is one and the folder's first page otherwise, so every folder with pages has one.

Author: ```suggestion
In 5.x a page was only considered a folder's "index" page only if its file was explicitly named `index`. The "index" page is now the page named `index` _when there is one_, falling back to the folder's first page otherwise.
```

Author: Applied, minus the "landing" framing, and with the doubled "only" fixed:

> In 5.x a page was only considered a folder's index page if its file was explicitly named `index`. The index page is now the page named `index` _when there is one_, falling back to the folder's first page otherwise.

*Drafted by Claude, reviewed before posting.*

> `:section` still yields `index`, and it is present for every folder now:

Author: ```suggestion
To ensure consistent behavior post-migration, use `x.section.title` instead of `x.index.page.title`. For example, the `<PageNav>` component `:section` block still yields `index`, but it is present for every folder now:
```

Author: Applied close to verbatim — "the `<PageNav>` component's `:section` block still yields `index`, but it is present for every folder now", followed by the diff showing the one change a consumer has to make.

*Drafted by Claude, reviewed before posting.*

```diff
  <:section as |x|>
    {{#if x.index}}
-     <x.index.Link>{{x.index.page.title}}</x.index.Link>
+     <x.index.Link>{{x.section.title}}</x.index.Link>
    {{else}}

>       {{x.section.title}}

Author: This seems pretty edge-cas-ey now. How often do we have a folder with no pages in it? Is this even valid?

Author: It is valid, but you are right that it is not worth a sentence in a migration guide, and it is no longer in one.

Valid because the manifest is built from a path list: a directory containing only a `meta.json`, or only sub-directories that are themselves empty, produces a `PageTree` with no pages under it. `tree.first` is then `undefined` and `getIndexPage` answers `undefined`.

So it is a real return value that a caller has to handle, which is why it is documented on `getIndexPage` and typed as `Page | undefined` — but the guide no longer narrates it. The only place it surfaces in `<PageNav />` is the `{{else}}` branch of `:section`, and the docs say that is the one folder that reaches it.

*Drafted by Claude, reviewed before posting.*

    {{/if}}
  </:section>
```

> **Take the heading text from `x.section.title`.** Left as `x.index.page.title`, a folder without a page named `index` shows its first page's title where 5.x showed the folder's name. The `{{else}}` branch is unreachable for any folder with pages.

Author: ```suggestion
**Note the change from `x.index.page.title` to `x.section.title`.** Left as `x.index.page.title`, a folder without a page named `index` will show its first page's title where in 5.x it showed the folder's name.

Also note that the `{{else}}` branch is unreachable for any folder with pages.
```

Author: Applied. Both halves are in — the `x.index.page.title` → `x.section.title` note with the consequence of leaving it, and the unreachable `{{else}}` branch.

I kept them as one paragraph rather than two; happy to split if you would rather they read as separate warnings.

*Drafted by Claude, reviewed before posting.*

With no `:section` block of your own, headings render `x.section.title` instead of the index page's `name` — which in 5.x was the literal string `index`.

> `getIndexPage` and `isIndex` are gone. Use `<PageNav />`'s `index`, or `indexPageFor(tree)` on the docs service. For "is this page named `index`", the check is `page.name === 'index'`.

Author: ```suggestion
`getIndexPage` and `isIndex` are removed. Use `<PageNav />`'s `index`, or `indexPageFor(tree)` on the docs service. For "is this page named `index`", the check is `page.name === 'index'`.
```

Is `page.name === 'index'` case-insensitive or does it need to be?

Author: Reworked in `d03bb51`, and the suggestion no longer applies as written since `getIndexPage` is back and `page.name === 'index'` is no longer the recommended check.

To answer the question you asked though: `name` is derived from the filename by the build, so the comparison is case-sensitive, and a file named `Index.md` would not match. I did not change that here — it is the same rule build-time sorting uses, and the two agreeing is the point of this PR. Making both case-insensitive is defensible but is a separate behavior change with its own migration note.

*Drafted by Claude, reviewed before posting.*

> ## A folder's index page sorts first regardless of extension

Author: This is a bugfix, NOT a breaking change. Remove from here and note in the description instead. (Note that I've edited the description)

Author: Done in `632b114` — removed from the guide, and the content is in the PR description under "Behavior changes not in the upgrade guide", noted as a bugfix rather than a migration step.

One thing I would have argued to keep, so you can overrule me knowingly: for a folder with a `.gjs.md`/`.gts.md` index and no `meta.json` `order`, `group.list[0]` changes, and with it the page that group's own URL resolves to. That is a URL an app could be linking to. It is in the description text, just not in the guide.

*Drafted by Claude, reviewed before posting.*

`index.md` has always been hoisted to the top of the folder holding it. `index.gjs.md` and `index.gts.md` were not: the build strips those extensions before sorting runs, so the test never matched. They now sort first too.

Two things move on any folder with a `.gjs.md` or `.gts.md` index and no `meta.json` `order`:

- The nav lists the index page first, where it used to appear in alphabetical position.
- `group.list[0]` becomes that index page, and with it the page a group's own URL resolves to.

A folder with a `meta.json` `order` is unaffected — an explicit index is hoisted before the order is applied, so it cannot be placed second.

To keep the old placement, give that folder a `meta.json` `order`. Build-time sorting also matches the node's name rather than its path now, so a **folder** named `index` sorts first among its siblings.

> ## Removed types

Author: Move this below "Folder and group URLs redirect instead of erroring"

Author: Moved, along with a broader reorder — "Removed types" had been sitting between the two index-page sections, which is what made this read as out of place.

Order is now: `Collection` → `PageTree`/`section`, a folder's index page falls back to its first page, folder and group URLs redirect, page titles honor the first heading, `<PageNav />` renders page links by title, removed types.

*Drafted by Claude, reviewed before posting.*

For example, the `<PageNav>` component's `:section` block still yields `index`, but it is present for every folder now:

```diff
  <:section as |x|>
    {{#if x.index}}
-     <x.index.Link>{{x.index.page.title}}</x.index.Link>
+     <x.index.Link>{{x.section.title}}</x.index.Link>
    {{else}}
      {{x.section.title}}
    {{/if}}
  </:section>
```

**Note the change from `x.index.page.title` to `x.section.title`.** Left as `x.index.page.title`, a folder without a page named `index` will show its first page's title where in 5.x it showed the folder's name. The `{{else}}` branch is also unreachable for any folder that has pages.

With no `:section` block of your own, headings render `x.section.title` instead of the index page's `name` — which in 5.x was the literal string `index`.

`getIndexPage(tree)` follows the same change, so it now answers for every folder with pages where it used to answer `undefined`. Two smaller shifts come with it: it can descend into a first child folder, and it matches the page actually named `index` rather than any path ending in `index` — `api-index.md` no longer counts.

> `isIndex` is removed. It asked whether a node was named `index`, which on its own does not answer anything a nav needs. If you were using it to decide what a folder's heading links to, that is `getIndexPage(tree)`; to decide whether listing a page repeats its folder's heading, that is `isRedundantWithHeading(folder, page)`.

Author: `isRedundantWithHeading` should probably only be true if the page is the index/landing page right?

Author: You are right, and it was a real bug — thank you. Fixed in `f6a0b39`.

Title-only meant a page that merely *happened* to share its folder's title was dropped from the nav even when the heading linked somewhere else. Concretely: a folder titled `Guides` by its `meta.json`, holding an index page titled `Overview` and a page titled `Guides`. The heading says "Guides" and links to `Overview`; the page titled `Guides` was hidden, and that list was the only place it appeared. Unreachable.

Both halves are needed though, not identity alone — identity alone would hide an index page whose title differs from its folder's, and that one is worth showing precisely because it says something the heading does not. So it is now: the page is the folder's index page **and** their titles match.

The unit fixtures that missed this used `pages: []`, which is not a tree any real folder has; they model real ones now, and the new case fails against the old implementation.

*Drafted by Claude, reviewed before posting.*

To assist in this migration, `<PageNav />` will `assert` in development when it is still given a `:collection` block to remind you to migrate to `:section`. The assertion is stripped from production builds.

`Node` is `Page | PageTree`. The `Runtime/utilities/collection-utils` page is now `page-tree-utils`, with a redirect from the old URL.

## Index pages and landing pages

> A folder now has two distinct things, where 5.x conflated them:

Author: ```suggestion
A folder now has two distinct concepts, where 5.x conflated them:
```

Author: Not taken, because the two concepts are gone rather than clarified — that was the outcome of the later thread where you pushed back on the split existing in public API at all.

A folder has one index page now: the page named `index` when there is one, its first page otherwise. The section says that in one sentence instead of defining a taxonomy first.

*Drafted by Claude, reviewed before posting.*

> - its **index page** — a page named `index`. The folder's own page. It may not have one.

Author: ```suggestion
- its explicit **index page** — a page named `index`. The folder's own page. It may not have one.
```

Author: Same as the sentence above it — the explicit/inclusive distinction is no longer public, so there is no bullet list of concepts to introduce.

Where the distinction still genuinely bites, it is stated locally rather than as vocabulary: a folder takes its *title* from a page actually named `index`, never from the first-page fallback, and the utilities page explains why (otherwise the fallback page would always be redundant with its own folder heading and vanish from the nav).

*Drafted by Claude, reviewed before posting.*

- its **landing page** — where the folder's URL goes: its index page if it has one, its first page otherwise. Every folder with pages has one.

> Most navigation wants the landing page, because it always exists.

Author: So why is the `isIndex` concept public API still?

Author: It is not — `isIndex` is removed, and stays removed. Checking whether a node is named `index` does not on its own answer anything a nav asks.

Worth recording that it *was* pre-existing published API: `kolay@5.4.0`'s `declarations/browser/index.d.ts` exports it, added by @NullVoxPopuli in `81790f6` (Dec 26 2025). So this is a real breaking removal, not the retraction of something we added on this branch, and the guide documents it as one with both replacements named (`getIndexPage` for the heading's target, `isRedundantWithHeading` for the redundancy check).

*Drafted by Claude, reviewed before posting.*

> ### `getIndexPage` is removed

Author: Was this documented public API or just something used under the hood?

Author: Documented public API, confirmed rather than assumed: `kolay@5.4.0`'s published `declarations/browser/index.d.ts` has `export { getIndexPage, isCollection, isIndex } from './utils.ts';`, and `getIndexPage` carried the "useful for making folder names in navigation link to an index page" doc comment that this PR keeps citing.

So it needed the guide entry. It is no longer a removal though — `d03bb51` restores it with the inclusive behavior, so the guide entry is now a behavior change.

*Drafted by Claude, reviewed before posting.*

> It returned the index page but was documented for the landing question — "useful for making folder names in navigation link to an index page" — so folders without an index page went unlinked.

Author: ```suggestion
It returned the index page but was documented for the landing question — "useful for making folder names in navigation link to an index page" — so folders without an explicit index page would not be linked.
```

Author: The quote you were preserving is the right one, and it is why `getIndexPage` came back rather than staying removed.

It was documented for "making folder names in navigation link to an index page" while returning only an explicit index — so index-less folders went unlinked. Making the definition inclusive makes the name and the documented purpose agree, which is a better outcome than deleting it.

So the guide now describes a behavior change: it answers for every folder with pages where it used to answer `undefined`.

*Drafted by Claude, reviewed before posting.*

> For a folder heading, use `<PageNav />`'s `landing`, or `landingForTree(tree)` on the docs service. For the narrow "does this folder have its own page" question, `isIndex` still answers it:

Author: What's the use-case for this logic?

Author: There was not a good one — that paragraph is deleted. It was explaining a distinction the public API no longer draws.

If you meant the underlying capability rather than the prose: the use case is a nav deciding what a folder's heading should link to, and that is `getIndexPage(tree)`, which the docs now show directly.

*Drafted by Claude, reviewed before posting.*

```diff
- const index = getIndexPage(folder);
+ const first = folder.pages.at(0);
+ const index = first && isIndex(first) ? first : undefined;
```

> Sorting puts an explicit index first in its folder, so checking the first child is enough.

Author: Be more specific about "sorting". Build-time sorting?

Author: Build-time, yes — said explicitly now. It matters because it is the reason the old rule was broken: the build strips `.gjs.md`/`.gts.md` before sorting runs, so a test against the *path* never matched those, and sorting had to move to the node's name.

That text has since moved to the PR description per your other comment, with "build-time" intact.

*Drafted by Claude, reviewed before posting.*

> ### `isIndex` matches the name

Author: ```suggestion
### `isIndex` matches the filename
```

This is more of a bugfix (for something that is pretty edge-case-y IMO) than a breaking change IMO. Probably should just be documented in the PR description but not in the upgrade guide.

Author: Overtaken by events in the direction you wanted: `isIndex` is removed outright, so there is no "matches the filename" behavior left to document as either a bugfix or a breaking change.

The removal itself does stay in the guide, because it was published API — `kolay@5.4.0` exports it — with both replacements named. The edge-case path-matching detail is not there.

*Drafted by Claude, reviewed before posting.*

It tested the path, so `api-index.md` counted as an index page. It tests the name now, so only a page actually named `index` does. It is also a type guard (`x is Page`), which narrows an entry from `folder.pages`.

### `<PageNav />`'s `:section` yields `landing`, not `index`

```diff
  <:section as |x|>
-   {{#if x.index}}
-     <x.index.Link>{{x.section.title}}</x.index.Link>
+   {{#if x.landing}}
+     <x.landing.Link>{{x.section.title}}</x.landing.Link>
    {{else}}
      {{x.section.title}}
    {{/if}}
  </:section>
```

> `landing` is present for every folder with pages, where `index` was only present for folders holding an `index` page. An index page is still left out of the `:page` block, since the section heading stands in for it; a first page acting as the landing stays in the list, because it is a page in its own right.

Author: Note to self: check to see if we have a full `PageNav` example in these docs (add one if not) and verify that it's still up-to-date after these changes.

Author: Checked, and we did not have one — the page was a one-line description plus a `ComponentSignature`. Added in `db52dd2`.

It spells out the default rendering of both blocks as a starting point to modify, notes that `x.index` is present for every folder with pages (so the `{{else}}` branch is only reached by a folder with none), and that a page the heading already links to is left out of `:page`. There is also a short "Replacing it entirely" section pointing at `getIndexPage` and `isRedundantWithHeading`, since those are exactly what an ejected nav needs.

It is accurate as of `f6a0b39`, including the redundancy fix you just caught.

*Drafted by Claude, reviewed before posting.*

### Folder titles

> `PageTree` carries a `title`, so a section heading no longer has to be derived by the app. It resolves as: the folder's own `title` from its `meta.json`, then its index page's title, then its cleaned directory name.

Author: Are all of these permutations accounted for in the table below? e.g. the case where we use h1 from the index?

Author: They were not, and checking turned up a genuine gap rather than a missing table row.

The h1 case you asked about only worked for `.gjs.md`/`.gts.md` pages, whose text the build inlines. A plain `.md` page was never read at build time, so it never got past its filename — the table was describing behavior most pages did not have. The build reads every markdown page's headings now, so the row is true as written.

The folder row needed the opposite treatment: it is the title of a page named `index` specifically, never the first-page fallback, and the table now says so with the reason underneath.

*Drafted by Claude, reviewed before posting.*

Page titles resolve the same way everywhere, navigation and search alike: the author's `title`, then the page's first heading, then the cleaned name.

If you pass no `:section` block, the default rendering changes:

| folder              | 5.x          | 6.0          |
| ------------------- | ------------ | ------------ |
| has an `index` page | `index`      | its title    |
| no `index` page     | `sub-folder` | `Sub folder` |

5.x rendered the index page's `name` — the literal string `index`. Set `title` in the folder's `meta.json` if the resolved one is not what you want.

## Removed types

`Options`, `MarkdownPagesOptions`, and `APIDocsOptions` (from `kolay/build` / `kolay/types`) described the old options shapes and are gone. `kolay/build` exports `DocsOptions` instead.

> ## A folder's index page now sorts first regardless of extension

Author: Is this actually a breaking change (needs migration guide) or a bugfix?

Author: Breaking, and it stayed — but a narrower version than what you were reading.

A folder with an `index` page rendered the literal string `index` as its heading in 5.x, because the default block used the page's `name`. It renders that page's title now. A folder without one rendered its raw directory segment and now renders the cleaned, capitalized form. Both are visible text changes in any app using the default rendering, which is what puts them in the guide rather than the description.

The genuinely bugfix-shaped item in that area — index sorting by extension — is out of the guide now, per your other comment.

*Drafted by Claude, reviewed before posting.*

`getIndexPage` keeps its name (it takes a `PageTree` now), and `Node` is `Page | PageTree`. The `Runtime/utilities/collection-utils` page is now `page-tree-utils`, with a redirect from the old URL.

## An index page and a landing page are now different things

Two words were doing one job, which is where a folder could end up with a heading pointing at one page while its URL went to another. They are now distinct throughout:

- an **index page** is a page explicitly named `index`. It is a folder's own page.
- a **landing page** is wherever a folder's URL goes: its index page when it has one, its first page otherwise.

> `isIndex` answers the first question and tested the path (`path.replace(/\.md$/, '').endsWith('index')`), so `api-index.md` counted. It now tests the name, through a shared `isIndexName` that sorting uses too — the two disagreed. It is also a type guard now (`x is Page`), so it narrows a `(Page | PageTree)[]` entry for you.

Author: Reword to make it clear what `isIndex` does now and how that's changed. This wording is confusing.

Author: Moot in the end — `isIndex` is removed rather than reworded, and the section it was in has been rewritten around the single inclusive `index`.

The confusion you were pointing at was real and structural: the old text was trying to describe two similar-sounding concepts at once. There is one now, so there is nothing to disambiguate.

*Drafted by Claude, reviewed before posting.*

`getIndexPage` is **removed**. It was misnamed for what it did, and its own documentation recommended it for the other question: "useful for making folder names in navigation link to an index page" — which left folders without an index page unlinked. Replace it with whichever question you meant:

```diff
  // "does this folder have its own page?"
- const index = getIndexPage(node);
+ const first = node.pages.at(0);

> + const index = first && isIndex(first) ? first : undefined;

Author: This seems pretty hacky. I know we want to get rid of `getIndexPage` bc it was mis-named...what is the use case for needing this particular feature in public API?

Author: You talked me out of the removal. `getIndexPage` is restored in `d03bb51`, with the inclusive definition: the page named `index` when the folder has one, its first page otherwise.

The original objection was that the name was wrong for what it did — it answered "is there a page named `index`" while its own docs recommended it for "where should this folder's heading link". Making the definition inclusive resolves that by making the two questions the same question, so the name is now accurate rather than misnamed.

It also reads `tree.first` rather than re-deriving the answer, so it cannot disagree with where the folder's URL actually redirects — which is the class of bug this PR exists to close.

*Drafted by Claude, reviewed before posting.*

```

> Sorting always hoists an explicit index to the front of its folder, so the first child is it whenever there is one. For "where should this folder's heading link", use the nav's landing page instead — see below.

Author: ```suggestion
Sorting always hoists an explicit index to the front of its folder, so the first child is that explicit index page whenever if it exists. For "where should this folder's heading link", use the nav's landing page instead — see below.
```

Author: Superseded — that paragraph is gone with the index/landing split, so "use the nav's landing page instead" no longer has a referent.

The substance you were tightening did survive: sorting hoists an explicit index to the front of its folder, which is why `tree.first` and "the page named `index`" are the same answer. `getIndexPage` now reads `tree.first` directly rather than restating the rule, so the two cannot drift.

*Drafted by Claude, reviewed before posting.*

`<PageNav />`'s `:section` block yields `landing` where it used to yield `index`, because that is what it always was once a folder had no index page to offer. It is present for **every** folder with pages:

```diff
  <:section as |x|>
-   {{#if x.index}}
-     <x.index.Link>{{x.section.name}}</x.index.Link>
+   {{#if x.landing}}
+     <x.landing.Link>{{x.section.name}}</x.landing.Link>
    {{else}}
      {{x.section.name}}
    {{/if}}
  </:section>
```

> Renaming rather than redefining is deliberate: `x.index` kept working would have meant every existing template silently pointing somewhere new.

Author: Preston hates when we use the word "silently"

Author: Removed here, and everywhere else on the branch — one more had crept into a `docs-app` comment. Noted for future prose.

*Drafted by Claude, reviewed before posting.*

> An explicit index is still omitted from the `:page` block, since the heading stands in for it. A first page serving as the landing is not — it is a page in its own right and stays in the list.

Author: is this a "section heading"? what do we mean by heading here?

Author: "Section heading" was jargon leaking out of the component's block names. The docs say what it is instead: a nav renders a folder as a heading that links to the folder's index page, and the `:section` block is what produces it.

The helper that decides whether a page duplicates that heading is named for it now too — `isRedundantWithHeading` — so the word has a definition attached rather than being assumed.

*Drafted by Claude, reviewed before posting.*

On the docs service, `landingForTree(tree)` answers the landing question for a tree you already hold, and `landingForPageTree(path, group?)` resolves a path to a tree and delegates.

> ## Removed types

Author: Seems like a non-sequitor. Should the sections about index/landing pages be grouped together?

Author: Why didn't we move the sections around accordingly?

Author: It was, and they are now grouped. "Removed types" had wedged itself between the two index-page sections.

The guide runs: `Collection` is now `PageTree`/`section` → a folder's index page falls back to its first page → folder and group URLs redirect → page titles honor the first heading → `<PageNav />` renders page links by title → removed types. Index and URL behavior together, then titles together, with the unrelated housekeeping last.

"Why didn't we move the sections around accordingly" — no good reason; I had been appending.

*Drafted by Claude, reviewed before posting.*

`getIndexPage` keeps its name (it takes a `PageTree` now), and `Node` is `Page | PageTree`. The `Runtime/utilities/collection-utils` page is now `page-tree-utils`, with a redirect from the old URL.

## An index page and a landing page are now different things

> Two words were doing one job, which is where a folder could end up with a heading pointing at one page while its URL went to another. They are now distinct throughout:

Reviewer: Seems too clever and doesn't explain what someone needs to understand when migrating or working with this new feature

Author: Agreed, and it is gone. That whole section defined an "index page" against a "landing page" — a split that only existed to answer one question (does the folder's heading already say what this page's link would say). That question is now a named helper, `isRedundantWithHeading(folder, page)`, and the guide just documents the one change to `index`.

The section is now titled "A folder's index page falls back to its first page" and leads with what a consumer has to change, in `d03bb51`.

*Drafted by Claude, reviewed before posting.*

- A **folder** named `index` now sorts first among its siblings. Folders were never matched before, because a folder's path is a bare segment.
- A page whose basename merely _ends_ in `index` no longer hoists. `api-index.md` used to, because `/foo/api-index.md` satisfies `path.endsWith('index.md')`; now only a page actually named `index` is hoisted. Rename it, or give the folder a `meta.json` `order`.

## A page tree's URL resolves to its first page, without wiring

`/Group/sub-folder` used to render the error page. It now redirects to that tree's first page, wired by `setupKolay` to the router rather than by any hook you call.

Two consequences:

- A URL that used to error now navigates. Anything asserting on the error page for a folder URL will need updating.
- A group root resolves this way too, on every mount shape — so `/Group` on a top-level mount and `/guides` on a nested one no longer depend on `handlePotentialIndexVisit`. An app that never called it gets those redirects anyway. This also fixes the case no route hook could serve: arriving at a mount's own URL from a page already inside that mount, which is where the group's own nav link points.

> Keep calling `handlePotentialIndexVisit` for the app root (`/`), which names no group for a transition to resolve. Calling it elsewhere is harmless.

Reviewer: This seems more like a reasoning description for the changelog than phrasing that is actionable to consumers

Reviewer 2: claude -- stop putting temporal PR information in docs. ugh

Author: Both fair, and the sentence is deleted. It explained a decision made in this PR, which a consumer reading the guide next year has no stake in — that belongs in the PR, not the docs.

I have re-read the guide for the same failure elsewhere and did not find another instance; the remaining "5.x did X, now it does Y" phrasing is migration-relevant history rather than PR history. The section it sat in has since been rewritten anyway (`d03bb51`).

*Drafted by Claude, reviewed before posting.*
